### PR TITLE
fix(@aws-amplify/datastore): fixes observeQuery not removing newly-filtered items from snapshot

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

`observeQuery()` was not removing items from the snapshot if they later become invalid through an update. Those updates were being pre-filtered by an inner `observe`, leaving `observeQuery` with no signal to remove them from or emit a new snapshot.

#### Issue #, if available

#9325 #9573

#### Description of how you validated changes

New tests added that simulate two update-to-remove scenarios; one during generation of the initial snapshot(s), and another after sync is done.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/DataStore.ts [hunk 1]
```diff
@@ -294,6 +296,27 @@ describe('DataStore observeQuery, with fake-indexeddb and fake sync', () => {
...
+		//      ( cover your eyes )
+		//
+		// this prevents pollution between tests that may include observe() calls.
+		// This is a cheap solution let DataStore "settle" before clearing it and
+		// starting the next test. If we don't do this, we get "spooky"
+		// contamination between tests.
+		//
+		// NOTE: If you know of a better way to isolate these tests, please
+		// replace this pause!
+		//
```
Comment (manueliglesias) [inline packages/datastore/__tests__/DataStore.ts:312]: Can we avoid this naughtiness by using different models for each test?

It may make the test more verbose but will make them more isolated from each other

**Review by manueliglesias — CHANGES_REQUESTED** [review]
The logic looks good to me, I would love to have more robust test even if that means defining a model per test to prevent cross talk

File: packages/datastore/__tests__/DataStore.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:312]: I don't think that helps with the errors I was seeing. But, I'd be happy to pair on it if you like. As ugly and "naughty" as this feels, it *does* solve the problem, whereas separate models would be unscalable, IMO, and would not be expected to solve the issue I was seeing. (Internal DataStore/storage engine `undefined` refs -- I didn't note what they were specifically, but we can pair up and take a look!)

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -1205,41 +1217,83 @@ class DataStore {
...
+
 			(async () => {
 				try {
 					// first, query and return any locally-available records
 					(await this.query(model, criteria, sortOptions)).forEach(item =>
 						items.set(item.id, item)
 					);
 
-					// observe the model and send a stream of updates (debounced)
-					handle = this.observe(
```
Comment (david-mcafee) [inline packages/datastore/src/datastore/datastore.ts:1216]: According to the PR description: "observeQuery() was not removing items from the snapshot if they later become invalid through an update. Those updates were being pre-filtered by an inner observe, leaving observeQuery with no signal to remove them from or emit a new snapshot" - does this mean that `observe` is not filtering the items as we'd expect it to? i.e. is `observe` broken, or am I perhaps misunderstanding?

**Review by david-mcafee — COMMENTED** [review]
Left a question re: `observe`

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (svidgen) [reply packages/datastore/src/datastore/datastore.ts:1216]: Heh. It is confusing! The problem was that the criteria were being passed to `observe` and `observe` *was doing exactly what it's supposed to*.

The TLDR is that `observeQuery` was calling `observe` and passing the criteria through. So, if an item was in an initial result set, updated, and then *not* in the original set, `observe` would never tell `observeQuery` about it, because the updated item would not match `observe`'s give criteria.

In a little more detail, if it helps:

Old flow:

1. app calls `observeQuery(Model, m => m.x('eq', 'abc'))`
2. `observeQuery` calls `query(Model, m => m.x('eq', 'abc'))`
3. `observeQuery` calls `observe(Model, m => m.x('eq', 'abc'))`
4. app calls `DataStore.save(new Model({x: 'abc'}))`
5. `observe` sees it
6. `observe` checks to see if it matches (`m => m.x('eq', 'abc')`)
7. `observe` passes it to `observeQuery`
8. `observeQuery` adds it to the snapshot
9. app calls `DataStore.save(Model.copyOf(model, d => d.x = 'NOT abc!!!'))`
10. `observe` sees it
11. `observe` checks to see if it matches (`m => m.x('eq', 'abc')`)
12. **`observe` discards it**
13. `observeQuery` gets no update: **the snapshot now incorrectly still includes the outdated `{x: 'abc'}` item**

The new flow moves the criteria test.

1. app calls `observeQuery(Model, m => m.x('eq', 'abc'))`
2. `observeQuery` calls `query(Model, m => m.x('eq', 'abc'))`
3. `observeQuery` calls `observe(Model, Predicates.ALL)` (*effectively*)
4. app calls `DataStore.save(new Model({x: 'abc'}))`
5. `observe` sees it 
7. `observe` passes it to `observeQuery`
8. `observeQuery` checks to see if it matches `m => m.x('eq', 'abc')`
9. `observeQuery` adds it to the snapshot
10. app calls `DataStore.save(Model.copyOf(model, d => d.x = 'NOT abc!!!'))`
11. `observe` sees it
12. `observe` passes it to `observeQuery`
13. `observeQuery` checks to see if it matches `m => m.x('eq', 'abc')`
14. `observeQuery` **correctly removes the item from the snapshot**

File: packages/datastore/__tests__/DataStore.ts (hunk 1 above)
Comment (undefobj) [reply packages/datastore/__tests__/DataStore.ts:312]: Would a middle ground be to rely on the stop/start events rather than using the pause? I always find that sleep solutions bite us long term because you get into a non-deterministic state, which reduces psychological confidence levels from Engineers later when they run tests. You're basically introducing a type of isolation with the pause->stop->start workflow as a "reset" of sorts, with one of them being ambiguous. Could you just remove that piece and have an event driven sequence to check the state of the system to move on?

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (undefobj) [reply packages/datastore/src/datastore/datastore.ts:1216]: Reading this update what occurs to me is that the foundational data structure here which this change is managing bookkeeping is the "snapshot". Is this an in-memory array? While the above updated flow makes sense in the workflow of the app notifications to observers, it's a bit different from the Mutation queue in the outbox that `DataStore.save()` also writes to in this sequence. In the case of the outbox flow you have a CREATE followed by an UPDATE WITH CONDITION which means the original item will be written to a key and then the update will be appended to the end of the queue. This is to be expected as it gives us a causal chain of events. It's worth thinking about if we have two logic streams now - one publishing the event stream to the cloud, one publishing the subscriber stream to local observers - with differing logic and filtering rules, and if this will cause maintenance or logic race conditions in the future. Not enough to block the merge of this PR, but we should whiteboard this and look at the longer term ramifications.

File: packages/datastore/__tests__/DataStore.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts:312]: @undefobj Yeah, I was hoping the big banner would attract a fairly concrete "use this event" or "use this call pattern". And actually, as I look at this again, it also looks like I might have dropped a `stop()` at some point. That gives me a few things to poke at.

Comment (svidgen) [reply packages/datastore/__tests__/DataStore.ts]: Sure enough: I'm an idiot. I dropped a `stop()` from my `beforeEach`. Adding it back in appears to render the whole `afterEach` cleanup moot. It's been removed.

**Review by david-mcafee — APPROVED** [review]
LGTM! 🎉

**Review by manueliglesias — APPROVED** [review]
LGTM 👍
